### PR TITLE
test: run irqbalance tests serially to avoid race condition

### PR DESCRIPTION
**Pull request by sohankunkerkar**
Addresses the integration tests failure for https://github.com/cri-o/cri-o/issues/6751

#### What type of PR is this?

/kind ci
/kind flake

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: test/serial/irqbalance.bats [hunk 1]
```diff
@@ -1,7 +1,7 @@
 #!/usr/bin/env bats
 # vim: set syntax=sh:
 
-load helpers
+load ../helpers
```
Comment (haircommander) [inline test/serial/irqbalance.bats:4]: another option if this doesn't work: you could call `skip` in the `setup()` if an env variable isn't present and then set the env variable on the second bats call

Comment (haircommander) [reply test/serial/irqbalance.bats:4]: kind of like `TEST_USERNS` handling

File: test/test_runner.sh [hunk 2]
```diff
@@ -27,10 +27,13 @@ function execute() {
...
+# Serial tests
+SERIAL_TESTS=("${TESTS_ROOT[@]}/serial")
 
-# The number of parallel jobs to execute
+# The number of parallel jobs to execute non-serial tests
 export JOBS=${JOBS:-$(($(nproc --all) * 4))}
 
 # Run the tests.
-execute bats --jobs "$JOBS" --tap "${TESTS[@]}"
+execute bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
```
Comment (mrunalp) [inline test/test_runner.sh:38]: execute replaces the current binary with bats so we can't use bash syntax to combine with other invocation or split this out to the next line. Instead, can we just invoke bats without the exec?

Comment (sohankunkerkar) [reply test/test_runner.sh]: Fixed

File: test/serial/irqbalance.bats (hunk 1 above)
Comment (jlojosnegros) [reply test/serial/irqbalance.bats:4]: > another option if this doesn't work: you could call `skip` in the `setup()` if an env variable isn't present and then set the env variable on the second bats call

it works I tested it yesterday

File: test/test_runner.sh [hunk 3]
```diff
@@ -27,10 +27,13 @@ function execute() {
...
+# Serial tests
+SERIAL_TESTS=("${TESTS_ROOT[@]}/serial")
 
-# The number of parallel jobs to execute
+# The number of parallel jobs to execute non-serial tests
 export JOBS=${JOBS:-$(($(nproc --all) * 4))}
 
 # Run the tests.
-execute bats --jobs "$JOBS" --tap "${TESTS[@]}"
+bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
```
Comment (jlojosnegros) [inline test/test_runner.sh:38]: I think the return value of test_runner.sh should take into account the result of both test sets ... 

somethink like
```suggestion
# Run the tests.
bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
parallel_test_result=$?
bats --tap "${SERIAL_TESTS[@]}"
serial_test_result=$?

[ parallel_test_result && serial_test_result ]
```

File: test/test_runner.sh [hunk 4]
```diff
@@ -27,10 +27,13 @@ function execute() {
...
-TESTS=("${@:-.}")
+TESTS_ROOT=("${@:-.}")
+# Serial tests
+SERIAL_TESTS=("${TESTS_ROOT[@]}/serial")
 
-# The number of parallel jobs to execute
+# The number of parallel jobs to execute non-serial tests
 export JOBS=${JOBS:-$(($(nproc --all) * 4))}
 
 # Run the tests.
```
Comment (jlojosnegros) [inline test/test_runner.sh:37]: I think the return value of test_runner.sh should take into account the result of both test sets ... 

somethink like
```suggestion
# Run the tests.
bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
parallel_test_result=$?
bats --tap "${SERIAL_TESTS[@]}"
serial_test_result=$?

[ parallel_test_result && serial_test_result ]
```

File: test/test_runner.sh [hunk 5]
```diff
@@ -27,10 +27,13 @@ function execute() {
...
-execute bats --jobs "$JOBS" --tap "${TESTS[@]}"
+bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
+bats --tap "${SERIAL_TESTS[@]}"
```
Comment (jlojosnegros) [inline test/test_runner.sh:39]: I think the return value of test_runner.sh should take into account the result of both test sets ... 

somethink like
```suggestion
# Run the tests.
bats --jobs "$JOBS" --tap "${TESTS_ROOT[@]}"
parallel_test_result=$?
bats --tap "${SERIAL_TESTS[@]}"
serial_test_result=$?

[ parallel_test_result && serial_test_result ]
```

**Comment by jlojosnegros** [comment]
It seems `test_runner.sh` is run with the variable `TESTFLAGS` as parameter in some places. 
That variable sometimes contain `.bat` filenames to be executed, so this approach would not work as it is.
Right now `test_runner.sh` just handle its parameters as the test collection to execute which can be a single file, a list of files or a folder.

File: test/test_runner.sh [hunk 6]
```diff
@@ -21,16 +21,12 @@ fi
...
 TESTS=("${@:-.}")
 
-# The number of parallel jobs to execute
+# The number of parallel jobs to execute tests
 export JOBS=${JOBS:-$(($(nproc --all) * 4))}
 
 # Run the tests.
-execute bats --jobs "$JOBS" --tap "${TESTS[@]}"
+bats --jobs "$JOBS" --tap "${TESTS[@]}"
+TEST_SERIAL="Yes" bats --tap ./irqbalance.bats
```
Comment (haircommander) [inline test/test_runner.sh:32]: does calling `execute` twice not work?

Comment (sohankunkerkar) [reply test/test_runner.sh:32]: I haven't tried but just followed [this](https://github.com/cri-o/cri-o/pull/6777#discussion_r1153819894) comment

File: test/test_runner.sh (hunk 2 above)
Comment (haircommander) [reply test/test_runner.sh]: @mrunalp I believe `execute` was just a function we defined to run time and put the stderr into stdout, I don't think we need to drop it (though having two separate executes for the two  lines seems like a good idea)

**Comment by haircommander** [comment]
/approve

LGTM @cri-o/cri-o-maintainers PTAL

**Pull request merged**